wormhole-connect/src/routes/operator.ts

```diff
@@ -221,6 +225,114 @@ export default class RouteOperator {
   }
 }
 
+// This caches successful quote results from SDK routes and handles multiple concurrent
+// async functions asking for the same quote gracefully.
+//
+// If we are already fetching a quote and a second hook requests the same quote elsewhere,
+// we queue up a Promise in `QuoteCacheEntry.pending` that we resolve when the original
+// quote request is resolved. This just prevents us from making redundant API calls when
+// multiple components or hooks are interested in a quote.
+class QuoteCache {
+  ttl: number;
+  cache: Record<string, QuoteCacheEntry>;
+  pending: Record<string, QuotePromiseHandlers[]>;
+
+  constructor(ttl: number) {
+    this.ttl = ttl;
+    this.cache = {};
+    this.pending = {};
+  }
+
+  quoteParamsKey(routeName: string, params: QuoteParams): string {
+    return `${routeName}:${params.sourceChain}:${params.sourceToken}:${params.destChain}:${params.destToken}:${params.amount}`;
+  }
+
+  get(routeName: string, params: QuoteParams): QuoteResult | null {
+    const key = this.quoteParamsKey(routeName, params);
+    const cachedVal = this.cache[key];
+    if (cachedVal) {
+      if (cachedVal.age() < this.ttl) {
+        return cachedVal.result;
+      } else {
+        delete this.cache[key];
+      }
+    }
+
+    return null;
+  }
+
+  async fetch(
+    routeName: string,
+    params: QuoteParams,
+    route: SDKv2Route,
+  ): Promise<QuoteResult> {
+    const key = this.quoteParamsKey(routeName, params);
+    const pending = this.pending[key];
+    if (pending) {
+      // We already have a pending request for this key, so don't create a new one.
+      // Instead, subscribe to its result when it resolves
+      return new Promise((resolve, reject) => {
+        pending.push({ resolve, reject });
+      });
+    } else {
+      // We don't yet have a pending request for this key, so initiate one
+      route
+        .computeQuote(
+          params.amount,
+          params.sourceToken,
+          params.destToken,
+          params.sourceChain,
+          params.destChain,
+          { nativeGas: params.nativeGas },
+        )
+        .then((result: QuoteResult) => {
+          const pending = this.pending[key];
+          for (const { resolve } of pending) {
+            console.info(`resolving sub ${key}`);
+            resolve(result);
+          }
+          delete this.pending[key];
+
+          // Cache result
+          this.cache[key] = new QuoteCacheEntry(result);
+        })
+        .catch((err: any) => {
+          const pending = this.pending[key];
+          for (const { reject } of pending) {
+            reject(err);
+          }
+          delete this.pending[key];
+        });
+
+      // Initialize list of promises awaiting this result
```

Comment:
Should we create this promise and add it to `pending` before calling `route.computeQuote()` above in case `computeQuote` resolves before we reach here? I'm trying to think if there's a possible race condition

Reply:
yeah I can do that to be safe, good point. it seems unlikely but better to be defensive anyway.

Reply:
https://github.com/wormhole-foundation/wormhole-connect/pull/2535/commits/707a4815c84af583b16e600eff1fc5dda066afd4